File: src/python/logger/logger.py
import json
from datetime import datetime

LOG_FILE = "logs/system.log"

def log(level: str, source: str, message: str):
    log_entry = {
        "timestamp": datetime.timezone.utc().isoformat() + "Z",
        "level": level,
        "source": source,
        "message": message,
    }

    try:
        with open(LOG_FILE, "a") as log_file:
            log_file.write(json.dumps(log_entry) + "\n")
    except Exception as e:
        print(f"Error writing to log file: {e}")

File: src/cpp/logger/logger.cpp
#include "logger.hpp"
#include <fstream>
#include <ctime>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

void Logger::log(const std::string& level, const std::string& source, const std::string& message) {
    std::ofstream logFile("logs/system.log", std::ios::app);
    if (logFile.is_open()) {
        json logEntry;
        std::time_t now = std::time(nullptr);
        logEntry["timestamp"] = std::asctime(std::localtime(&now)); // ISO 8601 preferred
        logEntry["level"] = level;
        logEntry["source"] = source;
        logEntry["message"] = message;

        logFile << logEntry.dump() << std::endl;
        logFile.close();
    } else {
        std::cerr << "Error: Unable to open log file." << std::endl;
    }
}